sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/DataSource.scala

```diff
@@ -468,7 +489,13 @@ case class DataSource(
       throw new AnalysisException("Cannot save interval data type into external storage.")
     }
 
-    providingClass.newInstance() match {
+    val resolvedRelation = providingClass.newInstance() match {
+      case relationToCheck: DataSourceValidator =>
```

Comment:
ok this is a special case, let's leave the checking logic in `JDBCUtils`. But I think my proposal is still valid for external data source which can infer schema.